DNS servers for preconfigured networks

edwardstudy opened the pull request:
This PR worked on configuring DNS servers for preconfigured networks.

We merged PR 151: https://github.com/cloudfoundry/bosh-agent/pull/151#issuecomment-381711516 and modified this code https://github.com/cloudfoundry/bosh-agent/blob/c74db775183f3e0e68254279e649b802bd4bf63f/platform/net/ubuntu_net_manager.go#L100-L103
It skipped configuring network alias interfaces and caused empty virtual interfaces. We should continue to configure networked after writing DNS entries. 

@mikexuu Please help to view about network configs.

dpb587-pivotal commented:
Tracking in [Story #159703071](https://www.pivotaltracker.com/story/show/159703071)

jrussett commented:
Hi @edwardstudy,

We've reviewed the PR and think that it's going in the right direction, however, this change actually broke one of our integration tests when we ran it locally. Specifically, we saw the error:

```
• Failure [38.617 seconds]
DeleteARPEntries
/Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/delete_arp_entries_test.go:15
  on ubuntu
  /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/delete_arp_entries_test.go:133
    deletes ARP entries from the cache [It]
    /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/delete_arp_entries_test.go:134

    Timed out after 10.009s.
    Expected
        <string>:
    to equal
        <string>: <incomplete>

    /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/delete_arp_entries_test.go:141

    Full Stack Trace
        /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go:147 +0x47e
    github.com/cloudfoundry/bosh-agent/vendor/github.com/onsi/gomega/internal/asyncassertion.(*AsyncAssertion).Should(0xc00021ae80, 0x47c9140, 0xc00047add0, 0x0, 0x0, 0x0, 0xc00021ae80)
        /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go:48 +0x77
    github.com/cloudfoundry/bosh-agent/integration_test.glob..func5.7.1()
        /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/delete_arp_entries_test.go:141 +0x45f
    github.com/cloudfoundry/bosh-agent/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc0000c7f80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /Users/pivotal/workspace/src/github.com/cloudfoundry/bosh-agent/integration/integration_suite_test.go:35 +0xb2
    testing.tRunner(0xc000182300, 0x473b220)
        /usr/local/Cellar/go/1.11.1/libexec/src/testing/testing.go:827 +0x163
    created by testing.(*T).Run
        /usr/local/Cellar/go/1.11.1/libexec/src/testing/testing.go:878 +0x651
```

Could you please double check that all integration and unit tests pass and fix any errors if present?

Thanks,
@jrussett && @s4heid, CF BOSH

edwardstudy commented:
Hi, @jrussett.

I just ran an integration locally. 

```
bin/test-integration --provider=virtualbox
...
• [SLOW TEST:166.541 seconds]
sync_dns
/Volumes/journey/workspace/src/github.com/cloudfoundry/bosh-agent/integration/sync_dns_test.go:15
  does not skip verification if no checksum is sent
  /Volumes/journey/workspace/src/github.com/cloudfoundry/bosh-agent/integration/sync_dns_test.go:147
------------------------------

Ran 27 of 27 Specs in 4395.604 seconds
SUCCESS! -- 27 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 1h13m27.671743801s
Test Suite Passed
Cleaning up
==> default: Forcing shutdown of VM...
==> default: Destroying VM and associated drives...
```

Are you sure this failure was from our code?

s4heid commented:
Hey @edwardstudy, thanks for submitting this PR. We ran the tests again and can confirm they worked for us.

Thanks,
@pivotal-mp && @s4heid, CF BOSH

edwardstudy commented:
Hi, @s4heid. About `Preconfigured` issue, we want to remove `writeResolvConf` func which SoftLayer doesn't need. Because SoftLayer currently hands over configure networks(including `dns-nameservers` definition on `/etc/network/interfaces`) to agent. Agent will write `/etc/network/interfaces`.

I just set preconfigured as false on `/var/vcap/bosh/settings.json` manually and restarted agent and got correct DNS records on `/etc/resolv.conf`.

I'll raise a PR to remove `writeResolvConf` func and move `IsPreconfigured ` func to decide to skip `boot.platform.SetupNetworking(settings.Networks)`
on this section: https://github.com/cloudfoundry/bosh-agent/blob/719afe4619a86090b3240672e1cf7b56a290c0fa/agent/bootstrap.go#L98-L100.

I thought it makes sense and is a better solution for both. 

Do you guys agree to remove `writeResolvConf` func and relative changes? Because I found opensuse also write resolv.conf but didn't sure they are work because they only write `/etc/resolv.conf` simply: https://github.com/cloudfoundry/bosh-agent/blob/master/platform/net/opensuse_net_manager.go#L125-L156


Currently, I am changing code and test to proof.

Thanks.

edwardstudy commented:
I just tested with bosh-lite. It looked like that warden also need to configure nameservers because vm's network definition included DNS records:

```
{
  "networks": {
    "default": {
      "type": "manual",
      "ip": "10.244.0.2",
      "netmask": "255.255.255.0",
      "gateway": "10.244.0.1",
      "resolved": false,
      "use_dhcp": false,
      "default": [
        "dns",
        "gateway"
      ],
      "dns": [
        "8.8.8.8"
      ],
      "mac": "",
      "preconfigured": true
    }
  }
}
```

dpb587-pivotal commented:
@edwardstudy I'm confused on what this is trying to accomplish. If you want network configuration to be written, that includes the DNS servers which appear in the networks settings. Those network + DNS settings are [written](https://github.com/cloudfoundry/bosh-agent/blob/f4062d718e5a508fe555b775708247172578bb40/platform/net/ubuntu_net_manager.go#L388) to the interfaces configuration, and then it is the operating system's responsibility to update resolvconf files with the correct DNS settings from those files.

It sounds like you're suggesting the OS is not respecting whatever DNS settings are being configured in the network settings? If so, I think that's a separate issue and I would be hesitant on any agent changes until we see what is happening there.

Can you confirm if that is the scenario you're trying to solve? And investigate if or why DNS settings from the `interfaces` configurations are not propagating to your VMs resolvconf?

edwardstudy commented:
@dpb587-pivotal Yes, I agreed with the network configuration process should includes the DNS servers. But what I meant from the last comment is that I removed write resolv.conf function from preconfigured flag logic and tested it on bosh-lite. I just noticed that VM settings included DNS records and `preconfigured ` flag which means this VM should have DNS records and not overwrite interfaces file if I am correct. Is it what manifest desired? 

 The `preconfigured` flag means that whatever DNS settings is agent will skip configure DNS settings as long as `preconfigured` is **true**. If it was OK, I could raise a PR to remove writeResolvConf function and clean up preconfigured flag logic that what we want to do.

And I also changed [SL CPI](https://github.com/cloudfoundry/bosh-softlayer-cpi-release/commit/b7ca52b1a26be0d8544d29b3ad63652771571de0) to set the preconfigured flag as false about networks. It assumed that SL hand over all network configuration(including DNS) to bosh agent.

I ran BATs and it looked like a good result. So SL CPI will do not use preconfigured flag anymore.

About `It sounds like you're suggesting the OS is not respecting whatever DNS settings are being configured in the network settings?` I just didn't sure other infrastructures logic about preconfigure flag like bosh-lite then I suggested a preservation. I 100% agreed that preconfigure flag should be a clean meaning but worried about compatibility.

Thank you.
